pkg/kube/snapshot/snapshot.go

```diff
@@ -153,3 +157,32 @@ func NewSnapshotter(kubeCli kubernetes.Interface, dynCli dynamic.Interface) (Sna
 	}
 	return nil, errors.New("Snapshot resources not supported")
 }
+
+// We use regexp to match because errors written in vs.Status.Error.Message are strings
+// and we don't have any status code or other metadata in there.
+var transientErrorRegexp = regexp.MustCompile("the object has been modified; please apply your changes to the latest version and try again")
+
+// Use regexp to detect resource conflict error
+// If CSI snapshotter changes error reporting to use more structured errors,
+// we can improve this function to parse and recognise error codes or types.
+func isTransientError(err error) bool {
+	return transientErrorRegexp.MatchString(err.Error())
+}
+
+func waitOnReadyToUse(
+	ctx context.Context,
+	dynCli dynamic.Interface,
+	snapGVR schema.GroupVersionResource,
+	snapshotName,
+	namespace string,
+	isReadyFunc func(*unstructured.Unstructured) (bool, error),
+) error {
+	retries := 100
+	return poll.WaitWithRetries(ctx, retries, isTransientError, func(context.Context) (bool, error) {
```

Comment:
Question: Is this the right poll to use here? Why is 100 retries the correct value?

It appears to me that there is currently only 1 other file where Kanister code uses poll.WaitWithRetries.

It is more common to use `poll.Wait`, sometimes with the ctx set to have a timeout. Or to call `poll.WaitWithBackoff` and a maximum time set.

Reply:
We could do that. The difference here is that we only retry on this specific type of error, while `poll.WaitWithBackoff` would not distinguish that.
The `100` value is pretty arbitrary just to avoid the infinite wait. We could use `poll.WaitWithBackoffWithRetries` to configure max time to wait as well, but I don't know what's the good value in this context.

Current setting for the context timeout is infinity because there's a long-running process. Inserting a new time limit here would mean potentially creating a new failure scenario. Setting retries would worst case result in similar behaviour to what we have right now.

Reply:
As @pavannd1 indicated we can check this in as is and iterate on the polling mechanism-- this will improve robustness as is.

We can retry only on a specific error when using poll.Wait or poll.WaitWithBackoff by making the check for that error within the function called by poll. See for example how an `apierrors.IsNotFound` error is handled in RepoServiceHandler.createService.

See https://github.com/kanisterio/kanister/blob/2756ffb60874c5194df3ad1bc0f959806b223d5c/pkg/controllers/repositoryserver/handler.go#L138

Yes, there are some subtle differences between terminating a poll with a timeout or by retry count-- in particular whether ctx expiration can affect a call made within the poll loop. But I don't think those differences matter when the call being made is just a Get.

Ultimately the retry count is going to be picked based on a some knowledge of how long it is reasonable to wait for the snapshot controller to resolve the transient. In my mind it is clearer to just express that as a time value instead of trying to calculate the retry count from desired time and the backoff parameters.

See the many instances of `poll.Wait(timeoutCtx,...` in pkg/app where an app-specific timeout is known and used. Or pkg/kube.WaitForPodReady.

Reply:
Upon further reflection, I've realized that the code as implemented has the following behavior:
* Will wait indefinitely for the controller to update the status to be complete or some form of error.
* Once it encounters a transient error it will poll a bounded number of times (with default backoff parameters)
* If a non-transient error is encountered it will return an error immediately.

Importantly, setting a timeout on the context at beginning of all polling will affect the first wait, the wait for controller to ready snapshot before setting any error, transient or not.

While I still think it would be better to have an explicit time limit on retrying of transients, and we probably could work out a way to build it into a retry function closure, at that point it is no longer simple and clear.

Best solution is probably to keep retry count on trasients and add a comment about expected time that count corresponds to using default backoff.